packages/beacon-node/src/util/blobs.ts

```
@@ -110,3 +110,84 @@ export function computeDataColumnSidecars(
     };
   });
 }
+
+/**
+ * If the node obtains 50%+ of all the columns, it SHOULD reconstruct the full data matrix via the recover_matrix helper
+ * See https://github.com/ethereum/consensus-specs/blob/dev/specs/fulu/das-core.md#recover_matrix
+ */
+export async function recoverDataColumnSidecars(
+  partialSidecars: Map<number, fulu.DataColumnSidecar>
+): Promise<fulu.DataColumnSidecars | null> {
+  const columnCount = partialSidecars.size;
+  if (columnCount < NUMBER_OF_COLUMNS / 2) {
+    // We don't have enough columns to recover
+    return null;
+  }
+
+  if (columnCount === NUMBER_OF_COLUMNS) {
+    // full columns, no need to recover
+    return Array.from(partialSidecars.values());
+  }
+
+  const firstDataColumn = partialSidecars.values().next().value;
+  if (firstDataColumn == null) {
+    // should not happen because we check the size of the cache before this
+    throw new Error("No data column found in cache to recover from");
+  }
```

```ts
  // this function should never throw, we catched all errors inside
  const fullSidecars = await recover(partialSidecars);
```

The comment in the consumer expects no throws, should we just return `null` here and let the consumer return `RecoverResult.Failed`??

---

I changed the flow to let it throw all the way to `SeenGossipBlockInput` so it's not a concern anymore